### ISSUE TITLE
Google network calendar with ~4000 items takes ~8mn to load - lightning/calendar 68.1 - Provider for Google Calendar 68.1 (gdata-provider) - Thunderbird 68.1 ESR

### ISSUE THREAD

**Issue by richardleger** [kewisch/gdata-provider#14]
Steps to reproduce:

Following performance issues reported with CalDav network calendar in https://bugzilla.mozilla.org/show_bug.cgi?id=1502923 and https://bugzilla.mozilla.org/show_bug.cgi?id=1572823 (Work in Progress), I thought to take the same approach to measure performance and evaluate progress over time of loading Google Calendar with ~4000 items via Lightning/Calendar and Provider for Google Calendar (gdata-provider).

Environment setup steps:
    - Install Thunderbird 68.1 ESR (British version) from https://download-installer.cdn.mozilla.net/pub/thunderbird/releases/68.1.0/win32/en-GB/Thunderbird%20Setup%2068.1.0.exe
    - Create a new Thunderbird user profile
    - No email setup
    - Go to TB Menu > Add-ons > Extensions
    - Search for Provider for Google Calendar
    - Add this add-on via the add button
    - Restart Thunderbird
    - Set to show All Events in Find Events view (unifinder)
    - No columns selected in Find Events view (default settings)
    - Add one Google network calendar with ~4000 items

Actual results:
    - Loading the entire calendar takes more than 8 minutes!

Expected results:
    - Loading such calendar via .ics file over HTTP takes only 1 or 2 seconds max!

**Comment by richardleger** [comment]
Philipp,

Any advise on how loading speed of Google Calendar can be improved in Thunderbird 68.1 ESR?
Any setting tweaks available to end-users?

It take 2mn loading per 1000 items batch in Thunderbird lightning/calendar UI...

Is that related to bug https://bugzilla.mozilla.org/show_bug.cgi?id=501689 you refer to in https://bugzilla.mozilla.org/show_bug.cgi?id=1544320#c35 ?

Comparatively network wise (DevTools > Network) the situation is much better with Google network calendar than with CalDAV network calendar (as per bug https://bugzilla.mozilla.org/show_bug.cgi?id=1543953):

ADSL - All Events - No sorting (unifinder) - One Google calendar with ~4000items

|# |TB Version|LoadingTime|DevTools > Networking|
|:--|:--|:--|:--|
|1|**68.1.0** (32-bit) ESR|**~8m02s**|**6 Requests** - 4.17MB Transferred - Finish: 6m65s|

Regards,

**Comment by kewisch** [comment]
It is super difficult to make this bug actionable. I think I mentioned in one of the bugzilla bugs, can you create a profile using the gecko profiler? I'd like to figure out if there is something we need to speed up in the Provider or if this is a Lightning bug. I'm assuming the latter, to be honest.

**Comment by richardleger** [comment]
Here is the Thunderbird Dev Tools Performance performance profile you requested... while loading the first 2000 items (about) of my Google "home" calendar with about 4000 items in it.

[gdata.#14.devtools.perf.profile.tb.68.1.1.loading.2000first.items.zip](https://github.com/kewisch/gdata-provider/files/3674947/gdata.14.devtools.perf.profile.tb.68.1.1.loading.2000first.items.zip)

I observe loading progress via the status bar that show "Synchronising home event item X of Y" where X grows incrementally and Y by steps of 1000...

After **1mn23sec** from start when reaching **1971** items (about) there is a **long "pause"** with Thunderbird "flickering" possibly not responding but the mention does not appear... this "pause" last for about **26sec** before resuming... 

So to **load** the **2000 first items** it **takes** in total about **1mn50sec**...

As further loading progress, the loading slow down, especially pauses re-occurs at some interval causing **the Google calendar to take about 8mn to load entirely**.

The same calendar **takes only 3mn to load entirely via CalDAV** in TB 68.1.1 (half+ better than via gdata) which in itself is way too long, it has reported separately here... https://bugzilla.mozilla.org/show_bug.cgi?id=1572823#c3 ... Lightning performance is a work in progress...

So while Lightning is certainly part of the delay, gdata-provider may have its part to play as well... or interaction between the two... causing extra delay... I had opened a bugzilla bug on Mozilla for issue reported here, but it was closed arguing this was a gdata issue, though I am conscious this is not entirely true as Lightning perf also has an impact obviously :-)

I intent to re-run same tests, later on, with more recent version of TB... would gdata-provider add-ons be available for those 70.x or 71.x branches?

If there is a specific version of TB or gdata-provider you want me to run through let me know...

**Comment by richardleger** [comment]
> It is super difficult to make this bug actionable. I think I mentioned in one of the bugzilla bugs, can you create a profile using the gecko profiler? I'd like to figure out if there is something we need to speed up in the Provider or if this is a Lightning bug. I'm assuming the latter, to be honest.

Info provided in previous comment...

**Comment by richardleger** [comment]
Any update?

**Comment by kewisch** [comment]
I probably wont have time to look in to this any time soon, but if someone else wants to take a stab at it I'm happy to review. Sorry I can't give you better news!

**Comment by richardleger** [comment]
> I probably wont have time to look in to this any time soon, 

Ok. Too bad... but understandable... was worth a try... I did what I could to help...

> but if someone else wants to take a stab at it I'm happy to review.

May be worth to advertise this on the git hub within the project readme.md and also on the Thunderbird add-on page perhaps... and in tb-planning... perhaps...

If you or anyone take over in the future, please know that:

* comparatively the same calendar takes only 3mn to load via Lightning/Caldav in TB 68.x branch (Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=1572823#c3) so issue is definitively not only due to Lightning...

* it can easily be actioned/reproducible by setting up a ~4000items calendar on Google and try to load it first time via gdata-provider add-on...

At the meantime, if by any chance you are still able to help bring performance up in Lightning Caldav... please do so... as performance seems regressing again... see at https://bugzilla.mozilla.org/show_bug.cgi?id=1572823#c16 
:-)

Thanks for your help.

**Comment by Char-Aznable** [comment]
I am getting the same problem of slow calendar item loading

**Comment by oskar-addiator** [comment]
2020 says hello with the same problem. Major freezes after 7346 items of 8312 has been loaded.

**Comment by pryorda** [comment]
reporting this issue as well.

**Comment by dtorgy** [comment]
Reporting this issue as well. Loading Thunderbird the first time and refreshing calendar events both make Thunderbird unusable for several minutes at a time.

**Comment by kewisch** [comment]
Sorry if this was closed preemptively, I've just set up a github action intended to close newer issues with the needinfo label. That said, I don't think there is immediate action I can take here. Performance is definitely a concern, but there are limitations in what a provider can do. I've done some tweaks in the latest branch on main if someone wants to give it a go. In terms of filing issues, if there is a specific path of code that can be optimized I'm all for having an issue open, otherwise it will be difficult for me to fix the general case.
